File: pyjiffy/jiffy/storage/block_client.py
from thrift.Thrift import TApplicationException, TMessageType
from thrift.protocol.TBinaryProtocol import TBinaryProtocolAccelerated
from thrift.transport import TTransport, TSocket
from thrift.transport.TTransport import TTransportException

from jiffy.storage import block_request_service
from jiffy.storage import block_response_service


class ClientEntry:
    def __init__(self, transport, protocol, client):
        self.transport = transport
        self.protocol = protocol
        self.client = client


class BlockClientCache:
    def __init__(self, timeout_ms):
        # self.cache = {}
        self.timeout_ms = timeout_ms

    def __del__(self):
        pass
        # for k in self.cache:
        #     client_entry = self.cache[k]
        #     if client_entry.transport.isOpen():
        #         client_entry.transport.close()

    def remove(self, host, port):
        pass
        # if (host, port) in self.cache:
        #     if self.cache[(host, port)].transport.isOpen():
        #         self.cache[(host, port)].transport.close()
        #     del self.cache[(host, port)]

    def get(self, host, port):
        # if (host, port) in self.cache:
        #     entry = self.cache[(host, port)]
        #     return entry.transport, entry.protocol, entry.client
        socket = TSocket.TSocket(host, port)
        socket.setTimeout(self.timeout_ms)
        transport = TTransport.TFramedTransport(socket)
        protocol = TBinaryProtocolAccelerated(transport)
        client = block_request_service.Client(protocol)

        ex = None
        for i in range(3):
            try:
                transport.open()
            except TTransportException as e:
                ex = e
                continue
            except Exception:
                raise
            else:
                break
        else:
            raise TTransportException(ex.type, "Connection failed {}:{}: {}".format(host, port, ex.message))
        # self.cache[(host, port)] = ClientEntry(transport, protocol, client)
        return transport, protocol, client


class CommandResponseReader:
    def __init__(self, iprot):
        self.iprot = iprot

    def recv_response(self):
        (fname, mtype, rseqid) = self.iprot.readMessageBegin()
        if mtype == TMessageType.EXCEPTION:
            x = TApplicationException()
            x.read(self.iprot)
            self.iprot.readMessageEnd()
            raise x
        args = block_response_service.response_args()
        args.read(self.iprot)
        self.iprot.readMessageEnd()
        if args.seq is not None and args.result is not None:
            return args.seq.client_seq_no, args.result
        raise TApplicationException(TApplicationException.MISSING_RESULT, "Command response failure: unknown result")

    def recv_responses(self, count):
        return [self.recv_response() for _ in range(count)]


class BlockClient:
    def __init__(self, client_cache, host, port, block_id, block_seq_no):
        self.transport_, self.protocol_, self.client_ = client_cache.get(host, port)
        self.id_ = block_id
        self.block_seq_no_ = block_seq_no

    def __del__(self):
        self.transport_.close()

    def is_open(self):
        return self.transport_.isOpen()

    def get_response_reader(self, client_id):
        self.client_.register_client_id(self.id_, client_id)
        return CommandResponseReader(self.protocol_)

    def get_client_id(self):
        return self.client_.get_client_id()

    def send_request(self, seq, arguments):
        self.client_.command_request(seq, self.id_, self.inject_seq_no(arguments))

    def send_run_command(self, block_id, arguments):
        self.client_.send_run_command(block_id, self.inject_seq_no(arguments))

    def recv_run_command(self):
        return self.client_.recv_run_command()

    def inject_seq_no(self, arguments):
        if len(arguments) == 0:
            return arguments[:]
        return [arguments[0], "$block_seq_no$", str(self.block_seq_no_)] + arguments[1:]

File: jiffy4j/core/src/main/java/jiffy/storage/BlockClient.java
package jiffy.storage;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import jiffy.storage.block_response_service.response_args;
import jiffy.util.ByteBufferUtils;
import org.apache.thrift.TApplicationException;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TMessage;
import org.apache.thrift.protocol.TMessageType;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

public class BlockClient {

  class CommandResponse {
    long clientSeqNo;
    public List<ByteBuffer> result;

    CommandResponse(long clientSeqNo, List<ByteBuffer> result) {
      this.clientSeqNo = clientSeqNo;
      this.result = result;
    }
  }

  class CommandResponseReader {

    private TProtocol protocol;

    CommandResponseReader(TProtocol protocol) {
      this.protocol = protocol;
    }

    CommandResponse receiveResponse() throws TException {
      TMessage message = protocol.readMessageBegin();
      if (message.type == TMessageType.EXCEPTION) {
        TApplicationException exception = new TApplicationException();
        exception.read(protocol);
        protocol.readMessageEnd();
        throw exception;
      }
      block_response_service.response_args args = new response_args();
      args.read(protocol);
      protocol.readMessageEnd();
      if (args.isSetSeq() && args.isSetResult()) {
        return new CommandResponse(args.getSeq().getClientSeqNo(), args.getResult());
      }
      throw new TApplicationException(TApplicationException.MISSING_RESULT,
          "Command response failure: unknown result");
    }
  }

  private int blockId;
  private Integer sequenceId;
  private TTransport transport;
  private TProtocol protocol;
  private block_request_service.Client client;

  BlockClient(BlockClientCache cache, String host, int port, int blockId, Integer sequenceId)
      throws TTransportException {
    BlockClientCache.Value value = cache.get(host, port);
    this.transport = value.getTransport();
    this.protocol = value.getProtocol();
    this.client = value.getClient();
    this.blockId = blockId;
    this.sequenceId = sequenceId;
  }

  public void close() {
    transport.close();
  }

  CommandResponseReader newCommandResponseReader(long clientId) throws TException {
    client.registerClientId(blockId, clientId);
    return new CommandResponseReader(protocol);
  }

  long getClientId() throws TException {
    return client.getClientId();
  }

  void sendCommandRequest(sequence_id seq, List<ByteBuffer> args) throws TException {
    System.out.println("sendCommand" + blockId);
    if(sequenceId != null) {
      System.out.println("seq:" + sequenceId);
      System.out.println("args:" + args.size());
      List<ByteBuffer> newArgs = new ArrayList<>(args);
      newArgs.add(ByteBufferUtils.fromString("seq:" + 1));
      args = newArgs;
    }
    client.commandRequest(seq, blockId, args);
  }
}